src/runtime_src/core/tools/xbtracer/src/lib/logger.h

```diff
@@ -0,0 +1,392 @@
+  const std::vector<std::pair<std::string, std::string>>& replacements);
+
+/*
+ * enumration to identify the log type (Entry/Exit)
+ * */
+enum class trace_type {
+  entry,
+  exit,
+  invalid
+};
+
+/*
+ * membuf class to dump the the memory buffer.
+ * */
+class membuf
+{
+  private:
+  unsigned char* m_ptr;
+  size_t m_sz;
+
+  public:
+  membuf(unsigned char* uptr, size_t sz)
+  {
+    m_ptr = uptr;
+    m_sz = sz;
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const membuf& mb)
+  {
+    for (unsigned int i = 0; i < mb.m_sz; i++)
+      os << std::to_string(*(mb.m_ptr + i)) << " ";
+    return os;
+  }
+
+  friend std::ofstream& operator<<(std::ofstream& ofs, const membuf& mb)
+  {
+    ofs.write("mem\0", 4);
+    ofs.write((const char*)&mb.m_sz, sizeof(mb.m_sz));
+    ofs.write((const char*)mb.m_ptr, mb.m_sz);
+    return ofs;
+  }
+};
+
+template <typename... Args>
+std::string stringify_args(const Args&... args);
+
+class logger
+{
+  private:
+  std::ofstream m_fp;
+  std::ofstream m_fp_bin;
+  std::string m_program_name;
+  bool m_inst_debug;
+  bool m_is_destructing = false;
+#ifdef _WIN32
+  DWORD m_pid;
+#else
+  pid_t m_pid;
+#endif /* #ifdef _WIN32 */
+  std::chrono::time_point<std::chrono::system_clock> m_start_time{};
+  std::thread synth_dtor_trace_thread;
+  std::vector<std::tuple<std::shared_ptr<xrt_core::device>, std::thread::id,
+                         std::string>> m_dev_ref_tracker;
+  std::vector<std::tuple<std::shared_ptr<kernel_impl>, std::thread::id,
+                         std::string>> m_krnl_ref_tracker;
+  std::vector<std::tuple<std::shared_ptr<run_impl>, std::thread::id,
+                         std::string>> m_run_ref_tracker;
+  std::vector<std::tuple<std::shared_ptr<bo_impl>, std::thread::id,
+                         std::string>> m_bo_ref_tracker;
+  std::vector<std::tuple<std::shared_ptr<hw_context_impl>, std::thread::id,
+                         std::string>> m_hw_cnxt_ref_tracker;
+
+  template <typename T>
+  bool check_ref_count(std::vector<std::tuple<std::shared_ptr<T>,
+                       std::thread::id, std::string>>& tuples)
+  {
+    bool bfound = false;
+  
+    for( auto it = tuples.begin(); it != tuples.end(); )
+    {
+      std::string dtor_name;
+      std::shared_ptr<T> pimpl;
+      std::thread::id tid;
+      std::tie(pimpl, tid, dtor_name) = *it;
+      auto count = pimpl.use_count();
+  
+      if (count > 2)
+      {
+        ++it;
+        bfound = true;
+      }
+      else
+      {
+        logger::get_instance().log(trace_type::entry, stringify_args(pimpl)
+                                   + "|" + dtor_name + "()|\n", tid);
+        logger::get_instance().log(trace_type::exit, stringify_args(pimpl)
+                                   + "|" + dtor_name + "||\n", tid);
+        tuples.erase(it);
+      }
+    }
+  
+    return bfound;
+  }
+
+  void synth_dtor_trace_fn();
+  
+  /*
+   * constructor
+   * */
+  logger();
+
+  public:
+  bool get_inst_debug()
```

Comment:
Just noticing now.  Make sure you declare member functions `const` when they can be.
```suggestion
  bool get_inst_debug() const
```

Reply:
Updated as suggested, please check.